File: mail.js
import { readFileSync } from 'node:fs';
import handlebars from 'handlebars';
import { convert } from 'html-to-text';
import nodemailer from 'nodemailer';
import { longDate, parseStart, shortDate, time } from './dates.js';

// Each milestone fires once per event, anywhere inside [from, to] days before the event.
// The window (rather than an exact day) is what lets a missed run catch up the next day;
// state.js guarantees it still only goes out once.
const MILESTONES = [
  { kind: 'announcement', template: 'mail_announcement', from: 26, to: 28 },
  { kind: 'invitation-7', template: 'mail_invitation', from: 5, to: 7 },
  { kind: 'invitation-2', template: 'mail_invitation', from: 1, to: 2 },
];

const templates = new Map();
const render = (name, data) => {
  if (!templates.has(name)) {
    const source = readFileSync(new URL(`./templates/${name}.hbs`, import.meta.url), 'utf-8');
    templates.set(name, handlebars.compile(source));
  }
  return templates.get(name)(data);
};

let transport;
const getTransport = (smtp) => {
  transport ??= nodemailer.createTransport(smtp);
  return transport;
};

/** Milestone due for this event today, or null. */
export const due = (ev, diff) => {
  if (ev.externalEvent) return null;
  const milestone = MILESTONES.find((m) => diff >= m.from && diff <= m.to);
  if (!milestone) return null;
  // The announcement doubles as the "registration is open" mail.
  if (milestone.kind === 'announcement' && ev.hideRegistration) return null;
  return milestone;
};

export const send = async (ev, milestone, config) => {
  const at = parseStart(ev.start);
  const html = render(milestone.template, { ...ev, date: longDate(at), time: time(at) });
  const prefix = milestone.kind === 'announcement' ? 'Ankündigung für ' : '';

  const info = await getTransport(config.smtp).sendMail({
    from: config.from,
    to: config.to,
    subject: `${prefix}${shortDate(at)}: ${ev.summary}`,
    html,
    // `strip` used to just delete the tags, which dropped every link from the text part.
    text: convert(html, { wordwrap: 78, selectors: [{ selector: 'a', options: { hideLinkHrefIfSameAsText: true } }] }),
  });
  return info.messageId;
};

File: src/Mail.java
import jakarta.mail.Authenticator;
import jakarta.mail.Message;
import jakarta.mail.PasswordAuthentication;
import jakarta.mail.Session;
import jakarta.mail.Transport;
import jakarta.mail.internet.InternetAddress;
import jakarta.mail.internet.MimeBodyPart;
import jakarta.mail.internet.MimeMessage;
import jakarta.mail.internet.MimeMultipart;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Properties;

public final class Mail {

  /**
   * Windows rather than exact days, so a dropped scheduled run catches up tomorrow.
   * These deliberately do not touch, which {@code MilestonesTest} asserts at the gaps.
   */
  private static final List<Milestone> MILESTONES = List.of(
      new Milestone("announcement", "mail_announcement", 26, 28),
      new Milestone("invitation-7", "mail_invitation", 5, 7),
      new Milestone("invitation-2", "mail_invitation", 1, 2));

  private Mail() {
  }

  /** The milestone due for this event today, if any. */
  public static Optional<Milestone> due(Event event, int daysUntil) {
    if (event.externalEvent()) {
      return Optional.empty();
    }
    return MILESTONES.stream()
        .filter(milestone -> milestone.covers(daysUntil))
        // The announcement doubles as the "registration is open" mail.
        .filter(milestone -> !(milestone.kind().equals("announcement") && event.hideRegistration()))
        .findFirst();
  }

  public static String render(Event event, Milestone milestone, LocalDateTime at) {
    return Templates.render(milestone.template(),
        event.model(Map.of("date", Dates.longDate(at), "time", Dates.time(at))));
  }

  public static String subject(Event event, Milestone milestone, LocalDateTime at) {
    var prefix = milestone.kind().equals("announcement") ? "Ankündigung für " : "";
    return prefix + Dates.shortDate(at) + ": " + event.summary();
  }

  public static String send(Event event, Milestone milestone, Config.Mail config) throws Exception {
    var at = Dates.parseStart(event.start()).orElseThrow();
    var html = render(event, milestone, at);

    var message = new MimeMessage(session(config.smtp()));
    message.setFrom(new InternetAddress(config.from()));
    message.setRecipients(Message.RecipientType.TO, InternetAddress.parse(config.to()));
    message.setSubject(subject(event, milestone, at), "UTF-8");

    var text = new MimeBodyPart();
    text.setText(Html.toPlainText(html), "UTF-8");
    var htmlPart = new MimeBodyPart();
    htmlPart.setContent(html, "text/html; charset=UTF-8");

    var body = new MimeMultipart("alternative");
    body.addBodyPart(text);
    body.addBodyPart(htmlPart);
    message.setContent(body);

    Transport.send(message);
    return message.getMessageID();
  }

  private static Session session(Config.Smtp smtp) {
    var properties = new Properties();
    properties.put("mail.transport.protocol", "smtp");
    properties.put("mail.smtp.host", smtp.host());
    properties.put("mail.smtp.port", String.valueOf(smtp.port()));
    if (smtp.secure()) {
      properties.put("mail.smtp.ssl.enable", "true");
    } else {
      // Upgrade when the server offers it; a submission port that does not is misconfigured.
      properties.put("mail.smtp.starttls.enable", "true");
    }
    if (!smtp.authenticates()) {
      return Session.getInstance(properties);
    }
    properties.put("mail.smtp.auth", "true");
    return Session.getInstance(properties, new Authenticator() {
      @Override
      protected PasswordAuthentication getPasswordAuthentication() {
        return new PasswordAuthentication(smtp.user(), smtp.password());
      }
    });
  }
}
